crates/task-impls/src/helpers.rs

```diff
@@ -60,6 +60,29 @@ pub async fn calculate_vid_disperse<TYPES: NodeType>(
     VidDisperse::from_membership(view, vid_disperse, membership.as_ref())
 }
 
+/// Calculate the vid disperse information from the payload given a view and membership, and precompute data from builder
+///
+/// # Panics
+/// Panics if the VID calculation fails, this should not happen.
+pub async fn calculate_vid_disperse_using_precompute_data<TYPES: NodeType>(
+    txns: Vec<u8>,
+    membership: Arc<TYPES::Membership>,
+    view: TYPES::Time,
+    pre_compute_data: VidPrecomputeData,
+) -> VidDisperse<TYPES> {
+    let num_nodes = membership.total_nodes();
+    let vid_disperse = spawn_blocking(move || {
+        #[allow(clippy::panic)]
+        vid_scheme(num_nodes).disperse_precompute(&txns, &pre_compute_data).unwrap_or_else(|err|panic!("VID disperse failure:\n\t(num_storage nodes,payload_byte_len)=({num_nodes},{})\n\terror: : {err}", txns.len()))
```

Comment:
Two things:
1. It looks like the formatter missed this, it seems like `|err|panic` should have a space.
2. I don't think it's a good idea to add new lines and tabs to error logs as it'll make it more difficult to parse when reading and applying structured logging.
3. You shouldn't need to panic here. Instead, you can remove `error: {err}` entirely and simply use `expect`.

Reply:
As an addendum, I'm not crystal clear why we spawn this as a blocking task? Can't you just call this would needing a background thread at all?

Reply:
I concur on the first to points. For the third... isn't `disperse_precompute` a bit long running, even with the precompute? I agree that `spawn_blocking` here is not helpful, but not on the reason. The advantage of `spawn` is that you can launch it allow it to run in a different thread, then come back and await (on the `JoinHandle`) after doing other stuff in parallel. It's a way to leverage the advantages of multithreaded concurrency in an async context. By contrast, `spawn_blocking` will yield on the calling async context while the spawned task gets to run (effectively) as if it were synchronous. This will allow other async tasks to progress, but not this current async task.

Crucially, either of these would only execute when you call `.await` on the future returned by `calculate_vid_disperse_using_precompute_data`, which isn't where the advantage of using `spawn` would be. To avoid serial execution of independent operations, you would want to `spawn` around `calculate_vid_disperse_using_precompute_data` and then await the result after completion of whatever other work needed doing in the calling context.

I suspect this is using `spawn_blocking` because of an (incorrect) rule of thumb that was widely promoted a few years ago, that recommended using it for longer (tens to hundreds of milliseconds) CPU-bound operations. This is still a one-off operation that eventually completes, and would be better handled with something like `rayon` or one of the more sophisticated threadpool crates.

Reply:
I guess it's fine since the existing api was following the same pattern. https://github.com/EspressoSystems/HotShot/blob/15c776d32001ce88edf2a87f3c520301da22a582/crates/task-impls/src/helpers.rs#L45